vllm-project/vllm
# [Benchmark] Benchmark structured output with datasets

Add structure output benchmark.

Base PR: https://github.com/vllm-project/vllm/pull/10046
Additional work:
1. add 4 guided options -  'grammar', 'choice', 'regex' and 'json'; For Json, add single-json and use 'xgrammar_benchmark'(multi-schema)
2. add option for guided_decoding_ratio, default is 1.0, ratio < 1.0 provided a mixed requests containing both regular/guided
3. add correctness rate check
4. add first and next token latency when testing with AsyncEngine

How to test:
* with guided decoding
```
python benchmarks/benchmark_guided.py --model meta-llama/Llama-3.2-3B-Instruct --dataset xgrammar_bench --async-engine --output-len 512 --num-prompts 10 --enable-chunked-prefill --guided-decoding-ratio 1.0 --save-results
```

How to test:
* with no guided decoding
```
python benchmarks/benchmark_guided.py --model meta-llama/Llama-3.2-3B-Instruct --dataset xgrammar_bench --output-len 512 --num-prompts 10 --no-guided-decoding --save-results
```

Expected output
FileName: 1.0guided_Llama-3.2-3B-Instruct_xgrammar_bench_10_out512_asyncTrue_warmupTrue_chunkedprefillTrue.txt

```
    "elapsed_time": 50.00640656100586,
    "num_requests": 10,
    "total_num_tokens": 8086,
    "total_output_tokens": 5120,
    "requests_per_second": 0.1999743770390778,
    "tokens_per_second": "161.70",
    "output_tokens_per_second": "102.39",
    "correct_rate(%)": 100.0,
    "first_token_latency(msecs)": {
        "count": 10.0,
        "mean": 33074.16370075662,
        "std": 9578.487238858703,
        "min": 19958.579740021378,
        "25%": 25430.36330281757,
        "50%": 33561.18999654427,
        "75%": 40371.71240762109,
        "max": 46179.445307003334
    },
    "next_token_latency(msecs)": {
        "count": 10.0,
        "mean": 7.487070074518177,
        "std": 0.21345235857034978,
        "min": 7.0583598064933035,
        "25%": 7.452180709461126,
        "50%": 7.483904802198513,
        "75%": 7.609307728370618,
        "max": 7.772003888715707
    }
```
